torch/csrc/autograd/engine.cpp

```diff
@@ -407,7 +407,12 @@ auto Engine::thread_main(const std::shared_ptr<GraphTask>& graph_task) -> void {
       }
 
       if (task.fn_ && !local_graph_task->has_error_.load()) {
+        // Set the ThreadLocalState before calling the function.
+        // NB: The ThreadLocalStateGuard doesn't set the grad_mode because GraphTask
+        // always saves ThreadLocalState without grad_mode.
```

Comment:
Noob question: what was the original reason for this comment - why can't we just have GraphTask save ThreadLocalState with grad_mode? Can we just have thread_locals_ in GraphTask keep_grad_mode=true now that we have both guards in the same place (maybe there are some instances that I am missing)?

Reply:
The gradmode is controlled by the `create_graph` flag passed to backward/grad. Not the current state of the thread.